File: src/seo_agents/week_spec.py
"""Frozen posting-week identity for the Friday SEO run.

GBP used to take date.today() as Day 1. A Saturday re-run therefore shifted
every GBP DATE by one day (2026-08-29 incident). Facebook already snaps to
Monday-on-or-after in fb_week_dates; this module is the single calendar both
crews and supabase-sync must use.

Rule (America/Chicago):
  run_friday = most recent Friday (Friday stays Friday; Sat/Sun rewind)
  week_of    = Monday on or after run_friday
  gbp_start  = run_friday  (keeps the live Fri–Thu GBP cadence)
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from datetime import date, datetime, timedelta, timezone
from pathlib import Path
from zoneinfo import ZoneInfo

CHICAGO = ZoneInfo("America/Chicago")
WEEK_SPEC_NAME = "week_spec.json"


@dataclass(frozen=True)
class WeekSpec:
    run_friday: str  # YYYY-MM-DD
    week_of: str  # YYYY-MM-DD Monday; seo_runs unique key
    gbp_start: str  # YYYY-MM-DD; GBP Day 1
    computed_at: str  # UTC instant

    def gbp_dates(self, days: int = 7) -> dict[int, date]:
        start = date.fromisoformat(self.gbp_start)
        return {d: start + timedelta(days=d - 1) for d in range(1, days + 1)}

    def to_dict(self) -> dict:
        return asdict(self)


def chicago_today(now: datetime | None = None) -> date:
    now = now or datetime.now(timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)
    return now.astimezone(CHICAGO).date()


def most_recent_friday(anchor: date) -> date:
    """Friday on or before ``anchor``. Monday→previous Friday, not next."""
    return anchor - timedelta(days=(anchor.weekday() - 4) % 7)


def monday_on_or_after(anchor: date) -> date:
    """Same rule as crew.fb_week_dates: a Monday anchor stays that Monday."""
    return anchor + timedelta(days=(7 - anchor.weekday()) % 7)


def compute_week_spec(anchor: date | None = None, now: datetime | None = None) -> WeekSpec:
    if anchor is None:
        anchor = chicago_today(now)
    run_friday = most_recent_friday(anchor)
    week_of = monday_on_or_after(run_friday)
    computed = (now or datetime.now(timezone.utc)).astimezone(timezone.utc).replace(microsecond=0)
    return WeekSpec(
        run_friday=run_friday.isoformat(),
        week_of=week_of.isoformat(),
        gbp_start=run_friday.isoformat(),
        computed_at=computed.isoformat().replace("+00:00", "Z"),
    )


def load_week_spec(output_dir: Path) -> WeekSpec | None:
    path = output_dir / WEEK_SPEC_NAME
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return WeekSpec(
            run_friday=data["run_friday"],
            week_of=data["week_of"],
            gbp_start=data.get("gbp_start") or data["run_friday"],
            computed_at=data.get("computed_at") or "",
        )
    except (OSError, KeyError, TypeError, json.JSONDecodeError, ValueError):
        return None


def save_week_spec(output_dir: Path, spec: WeekSpec) -> Path:
    output_dir.mkdir(parents=True, exist_ok=True)
    path = output_dir / WEEK_SPEC_NAME
    tmp = path.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(spec.to_dict(), indent=2) + "\n", encoding="utf-8")
    tmp.replace(path)
    return path


def load_or_create_week_spec(output_dir: Path, now: datetime | None = None) -> WeekSpec:
    """Reuse a same-Friday spec so a Saturday execute cannot shift dates.

    If the file is from a different Friday, recompute.
    """
    wanted = compute_week_spec(chicago_today(now), now)
    existing = load_week_spec(output_dir)
    if existing and existing.run_friday == wanted.run_friday:
        return existing
    save_week_spec(output_dir, wanted)
    return wanted

File: scripts/weekly/lib/week-spec.mjs
/**
 * week-spec.mjs — the frozen posting-week identity, ported from
 * src/seo_agents/week_spec.py and fb_week_dates in src/seo_agents/crew.py.
 *
 * Rule (America/Chicago):
 *   run_friday = most recent Friday on or before the anchor (Friday stays Friday;
 *                Saturday/Sunday/Monday rewind to the previous Friday)
 *   week_of    = Monday on or after run_friday (seo_runs unique key)
 *   gbp_start  = run_friday; GBP day N = gbp_start + (N - 1)   (Fri..Thu)
 *   fb_dates   = Mon (1), Wed (3), Fri (5), Sat (6) of the week_of week
 *
 * All date math is on YYYY-MM-DD strings via UTC epoch days so the host
 * timezone never leaks in; only "today" is derived from the Chicago clock.
 */
import { WeekSpecSchema } from './schemas.mjs';

export const CHICAGO_TZ = 'America/Chicago';
const DAY_MS = 86400000;
const FB_DAY_OFFSET = { 1: 0, 3: 2, 5: 4, 6: 5 };

const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

export function parseDate(s) {
  if (!DATE_RE.test(String(s))) throw new Error(`week-spec: not a YYYY-MM-DD date: ${s}`);
  const [y, m, d] = s.split('-').map(Number);
  const ms = Date.UTC(y, m - 1, d);
  const back = new Date(ms);
  if (back.getUTCFullYear() !== y || back.getUTCMonth() !== m - 1 || back.getUTCDate() !== d) {
    throw new Error(`week-spec: invalid calendar date: ${s}`);
  }
  return ms;
}

export function formatDate(ms) {
  return new Date(ms).toISOString().slice(0, 10);
}

export function addDays(s, days) {
  return formatDate(parseDate(s) + days * DAY_MS);
}

/** Python-style weekday: Monday = 0 ... Sunday = 6. */
export function weekday(s) {
  return (new Date(parseDate(s)).getUTCDay() + 6) % 7;
}

/** Today's calendar date in Chicago for the given instant. */
export function chicagoToday(now = new Date()) {
  const fmt = new Intl.DateTimeFormat('en-CA', {
    timeZone: CHICAGO_TZ, year: 'numeric', month: '2-digit', day: '2-digit',
  });
  const parts = Object.fromEntries(fmt.formatToParts(now).filter((p) => p.type !== 'literal').map((p) => [p.type, p.value]));
  return `${parts.year}-${parts.month}-${parts.day}`;
}

/** Friday on or before the anchor. Monday goes back to the previous Friday. */
export function mostRecentFriday(anchor) {
  const back = (((weekday(anchor) - 4) % 7) + 7) % 7;
  return addDays(anchor, -back);
}

/** Monday on or after the anchor; a Monday anchor stays that Monday. */
export function mondayOnOrAfter(anchor) {
  return addDays(anchor, (7 - weekday(anchor)) % 7);
}

export function gbpDates(gbpStart, days = 7) {
  const out = {};
  for (let d = 1; d <= days; d += 1) out[d] = addDays(gbpStart, d - 1);
  return out;
}

export function fbDates(weekOf) {
  const out = {};
  for (const [day, offset] of Object.entries(FB_DAY_OFFSET)) out[day] = addDays(weekOf, offset);
  return out;
}

/**
 * @param {object} [opts]
 * @param {string} [opts.anchor]  YYYY-MM-DD; defaults to Chicago today for `now`
 * @param {Date}   [opts.now]     instant used for the anchor default and computed_at
 */
export function computeWeekSpec({ anchor, now = new Date() } = {}) {
  const a = anchor || chicagoToday(now);
  const runFriday = mostRecentFriday(a);
  const weekOf = mondayOnOrAfter(runFriday);
  const spec = {
    run_friday: runFriday,
    week_of: weekOf,
    gbp_start: runFriday,
    computed_at: now.toISOString(),
    gbp_dates: gbpDates(runFriday),
    fb_dates: fbDates(weekOf),
  };
  const result = WeekSpecSchema.safeParse(spec);
  if (!result.success) throw new Error(`week-spec: internal error: ${result.error.issues[0].message}`);
  return result.data;
}

/** WeekSpec for a given week_of Monday (used when the wrapper passes --week-of). */
export function weekSpecForWeekOf(weekOf, now = new Date()) {
  if (weekday(weekOf) !== 0) throw new Error(`week-spec: --week-of must be a Monday, got ${weekOf}`);
  // The run Friday for a Monday week_of is the Friday three days before it.
  return computeWeekSpec({ anchor: addDays(weekOf, -3), now });
}
